sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -1583,21 +1599,20 @@ class Analyzer(override val catalogManager: CatalogManager)
     def resolveAssignments(
         assignments: Seq[Assignment],
         mergeInto: MergeIntoTable,
-        resolveValuesWithSourceOnly: Boolean): Seq[Assignment] = {
+        resolveValuesFrom: LogicalPlan): Seq[Assignment] = {
       assignments.map { assign =>
         val resolvedKey = assign.key match {
           case c if !c.resolved =>
             resolveMergeExprOrFail(c, Project(Nil, mergeInto.targetTable))
           case o => o
         }
         val resolvedValue = assign.value match {
-          // The update values may contain target and/or source references.
           case c if !c.resolved =>
-            if (resolveValuesWithSourceOnly) {
-              resolveMergeExprOrFail(c, Project(Nil, mergeInto.sourceTable))
-            } else {
-              resolveMergeExprOrFail(c, mergeInto)
+            val resolveFromChildren = resolveValuesFrom match {
+              case m: MergeIntoTable => m
+              case p => Project(Nil, p)
```

Comment:
this pattern match looks a bit hacky. How about adding a new method?
```
def resolveAssignmentsByPlanOutput ... {
  resolveAssignments(..., Project(Nil, p))
}
```

Reply:
or we pass an enum:
```
  object MergeResolvePolicy extends Enumeration {
    val BOTH, TARGET, SOURCE = Value
  }
```
then the code here can be
```
val basePlan = policy match {
  case BOTH => m
  case SOURCE => Project(Nil, m.source)
  case TARGET => Project(Nil, m.target)
}
```

Reply:
I went the enum route, introduced MergeResolvePolicy.